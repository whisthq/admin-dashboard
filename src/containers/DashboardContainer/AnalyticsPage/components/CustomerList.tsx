import React from 'react'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCircleNotch } from '@fortawesome/free-solid-svg-icons'

import '../../../../static/App.css'
import Style from '../../../../styles/components/pageAdmin.module.css'

import { fetchCustomers } from '../../../../actions/index'

class CustomerList extends React.Component<any, any> {
    constructor(props: any) {
        super(props)
        this.state = {
            customers_fetched: false,
        }
    }

    // intervalID var to keep track of auto-refreshing across functions
    intervalID: any | undefined

    componentDidMount() {
        // refresh the customer table every 60 seconds
        this.intervalID = setInterval(this.getUpdatedDatabase.bind(this), 60000)
        this.props.dispatch(fetchCustomers())
    }

    componentWillUnmount() {
        // stop auto-refreshing
        clearInterval(this.intervalID)
    }

    componentDidUpdate() {
        if (
            this.props.access_token &&
            this.props.customers.length === 0 &&
            !this.state.customers_fetched
        ) {
            this.props.dispatch(fetchCustomers())
            this.setState({ customers_fetched: true })

            // refresh the customer table every 60 seconds
            this.intervalID = setInterval(
                this.getUpdatedDatabase.bind(this),
                60000
            )
        }
    }

    getUpdatedDatabase() {
        this.props.dispatch(fetchCustomers())
    }

    render() {
        let component = this

        return (
            <div
                style={{
                    maxHeight: 300,
                    overflowY: 'scroll',
                }}
            >
                {this.props.customers.length > 0 ? (
                    <div
                        style={{
                            display: 'flex',
                            flexWrap: 'wrap',
                        }}
                    >
                        {this.props.customers.map(function (
                            value: any,
                            index: any
                        ) {
                            if (value && value['username'] !== '') {
                                return (
                                    <div
                                        key={index}
                                        className="pointerOnHover"
                                        onClick={() => {
                                            component.props.openModal(
                                                value['username']
                                            )
                                        }}
                                        style={{
                                            padding: '5px 10px',
                                            marginRight: 10,
                                            marginBottom: 10,
                                            fontSize: 13,
                                            color: '#8884d8',
                                            background:
                                                'rgba(136, 132, 216, 0.1)',
                                        }}
                                    >
                                        {value['username']}
                                    </div>
                                )
                            } else {
                                return <div></div>
                            }
                        })}
                    </div>
                ) : (
                    <div className={Style.spinnerContainer}>
                        <div style={{ width: '100%', textAlign: 'center' }}>
                            <FontAwesomeIcon
                                icon={faCircleNotch}
                                spin
                                style={{
                                    color: '#1e1f36',
                                    margin: 'auto',
                                    marginTop: 220,
                                }}
                            />
                        </div>
                    </div>
                )}
            </div>
        )
    }
}

function mapStateToProps(state: any) {
    return {
        customers: state.AccountReducer.customers
            ? state.AccountReducer.customers.reverse()
            : [],
        access_token: state.AccountReducer.access_token,
    }
}

export default connect(mapStateToProps)(CustomerList)
