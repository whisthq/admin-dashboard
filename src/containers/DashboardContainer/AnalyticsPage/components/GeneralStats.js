import React, { Component } from 'react'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCircleNotch } from '@fortawesome/free-solid-svg-icons'
import moment from 'moment'
import { ToggleButtonGroup, ToggleButton } from 'react-bootstrap'
import {
    LineChart,
    Line,
    ResponsiveContainer,
    XAxis,
    YAxis,
    Tooltip,
} from 'recharts'

import '../../../../static/App.css'

import { fetchRegionReport, fetchLatestReport } from '../../../../actions/index'

class GeneralStats extends Component {
    constructor(props) {
        super(props)
        this.state = {
            total: null,
            eastus: null,
            northcentralus: null,
            southcentralus: null,
            timescale: 'day',
        }
    }

    componentDidMount() {
        this.props.dispatch(fetchLatestReport())
        this.props.dispatch(fetchRegionReport(this.state.timescale))
    }

    componentDidUpdate(prevProps) {
        if (prevProps.regionReport !== this.props.regionReport) {
            let total = [],
                eastus = [],
                northcentralus = [],
                southcentralus = []

            this.props.regionReport.forEach((element) => {
                total.push({
                    x: element.timestamp,
                    'Number of users': element.users_online,
                })
                eastus.push({
                    x: element.timestamp,
                    'Number of users': element.eastus_unavailable,
                })
                northcentralus.push({
                    x: element.timestamp,
                    'Number of users': element.northcentralus_unavailable,
                })
                southcentralus.push({
                    x: element.timestamp,
                    'Number of users': element.southcentralus_unavailable,
                })
            })
            this.setState({
                total: total,
                eastus: eastus,
                northcentralus: northcentralus,
                southcentralus: southcentralus,
            })
        }
    }

    handleChartSelect = (val) => {
        this.setState({ timescale: val })
        this.props.dispatch(fetchRegionReport(val))
    }

    chartElement = (userType) => {
        let dataList = null
        switch (userType) {
            case 'total':
                dataList = this.state.total
                break
            case 'eastus':
                dataList = this.state.eastus
                break
            case 'northcentralus':
                dataList = this.state.northcentralus
                break
            case 'southcentralus':
                dataList = this.state.southcentralus
                break
            default:
                dataList = this.state.total
        }
        let format = 'MMM Do'
        if (this.state.timescale === 'day') {
            format = 'h A'
        } else {
            format = 'MMM Do'
        }

        return (
            <ResponsiveContainer height={150} width="100%">
                <LineChart data={dataList} syncId="usageChart">
                    <Line
                        type="monotone"
                        dataKey="Number of users"
                        dot={false}
                        strokeWidth={2}
                        stroke="#4636a6"
                    />
                    <XAxis
                        dataKey="x"
                        axisLine={false}
                        tickLine={false}
                        tick={{
                            fontSize: 12,
                            transform: 'translate(0, 10)',
                        }}
                        domain={['auto', 'auto']}
                        tickFormatter={(unixTime) =>
                            moment(unixTime * 1000).format(format)
                        }
                        type="number"
                        scale="time"
                    />
                    <YAxis
                        dataKey="Number of users"
                        axisLine={false}
                        tickLine={false}
                        allowDecimals={false}
                        tick={{
                            fontSize: 12,
                        }}
                    />
                    <Tooltip
                        contentStyle={{
                            border: 'none',
                            fontSize: 14,
                        }}
                        labelFormatter={(unixTime) =>
                            moment(unixTime * 1000).format(format)
                        }
                    />
                </LineChart>
            </ResponsiveContainer>
        )
    }

    render() {
        if (this.state.total) {
            return (
                <div>
                    <div className=" d-flex justify-content-between">
                        <p
                            style={{
                                fontWeight: 'bold',
                                fontSize: 24,
                            }}
                        >
                            Active Users
                        </p>
                        <div>
                            <ToggleButtonGroup
                                type="radio"
                                name="select"
                                defaultValue={'day'}
                                onChange={this.handleChartSelect}
                            >
                                <ToggleButton value={'day'}>
                                    24 Hours
                                </ToggleButton>
                                <ToggleButton value={'week'}>
                                    7 Days
                                </ToggleButton>
                                <ToggleButton value={'month'}>
                                    30 Days
                                </ToggleButton>
                                {/* <ToggleButton value={'all'}>
                                    All Time
                                </ToggleButton> */}
                            </ToggleButtonGroup>
                        </div>
                    </div>
                    <p className="chart-header">Total</p>
                    <div
                        style={{
                            position: 'relative',
                            right: 40,
                        }}
                    >
                        {this.chartElement('total')}
                    </div>
                    <p className="chart-header">Eastus</p>
                    <div
                        style={{
                            position: 'relative',
                            right: 40,
                        }}
                    >
                        {this.chartElement('eastus')}
                    </div>
                    <p className="chart-header">Northcentralus</p>
                    <div
                        style={{
                            position: 'relative',
                            right: 40,
                        }}
                    >
                        {this.chartElement('northcentralus')}
                    </div>
                    <p className="chart-header">Southcentralus</p>
                    <div
                        style={{
                            position: 'relative',
                            right: 40,
                        }}
                    >
                        {this.chartElement('southcentralus')}
                    </div>
                </div>
            )
        } else {
            return (
                <div>
                    <FontAwesomeIcon icon={faCircleNotch} spin />
                </div>
            )
        }
    }
}

function mapStateToProps(state) {
    return {
        latestReport: state.AccountReducer.latestReport,
        regionReport: state.AccountReducer.regionReport,
    }
}

export default connect(mapStateToProps)(GeneralStats)
