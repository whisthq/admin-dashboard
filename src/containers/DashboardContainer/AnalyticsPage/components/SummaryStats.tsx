import React from 'react'
import { connect } from 'react-redux'
import { Card, Row, Col } from 'antd'

import 'antd/dist/antd.css'

import { fetchTotalSignups, fetchTotalMinutes } from '../../../../actions/index'

class SummaryStats extends React.Component<any, any> {
    componentDidMount() {
        this.props.dispatch(fetchTotalSignups())
        this.props.dispatch(fetchTotalMinutes())
    }

    render() {
        return (
            <div>
                <p
                    style={{
                        fontWeight: 'bold',
                        fontSize: 24,
                    }}
                >
                    Summary
                </p>
                <Row gutter={16}>
                    <Col span={8}>
                        <Card
                            title="Total Session length"
                            size="small"
                            loading={this.props.totalMinutes == null}
                        >
                            {this.props.totalMinutes !== null && (
                                <div>
                                    <p>
                                        Past day:{' '}
                                        <strong>
                                            {Math.round(
                                                this.props.totalMinutes['day'] /
                                                    60
                                            ).toString() +
                                                'h ' +
                                                (
                                                    this.props.totalMinutes[
                                                        'day'
                                                    ] % 60
                                                ).toString() +
                                                'm'}
                                        </strong>
                                    </p>
                                    <p>
                                        Past week:{' '}
                                        <strong>
                                            {Math.round(
                                                this.props.totalMinutes[
                                                    'week'
                                                ] / 60
                                            ).toString() +
                                                'h ' +
                                                (
                                                    this.props.totalMinutes[
                                                        'week'
                                                    ] % 60
                                                ).toString() +
                                                'm'}
                                        </strong>
                                    </p>
                                    <p>
                                        Past month:{' '}
                                        <strong>
                                            {Math.round(
                                                this.props.totalMinutes[
                                                    'month'
                                                ] / 60
                                            ).toString() +
                                                'h ' +
                                                (
                                                    this.props.totalMinutes[
                                                        'month'
                                                    ] % 60
                                                ).toString() +
                                                'm'}
                                        </strong>
                                    </p>
                                </div>
                            )}
                        </Card>
                    </Col>
                    <Col span={8}>
                        <Card
                            title="Number of signups"
                            size="small"
                            loading={this.props.totalSignups == null}
                        >
                            {this.props.totalSignups !== null && (
                                <div>
                                    <p>
                                        Past day:{' '}
                                        <strong>
                                            {this.props.totalSignups['day']}
                                        </strong>
                                    </p>
                                    <p>
                                        Past week:{' '}
                                        <strong>
                                            {this.props.totalSignups['week']}
                                        </strong>
                                    </p>
                                    <p>
                                        Past month:{' '}
                                        <strong>
                                            {this.props.totalSignups['month']}
                                        </strong>
                                    </p>
                                </div>
                            )}
                        </Card>
                    </Col>
                </Row>
            </div>
        )
    }
}

function mapStateToProps(state: any) {
    return {
        totalSignups: state.AccountReducer.totalSignups,
        totalMinutes: state.AccountReducer.totalMinutes,
    }
}

export default connect(mapStateToProps)(SummaryStats)
