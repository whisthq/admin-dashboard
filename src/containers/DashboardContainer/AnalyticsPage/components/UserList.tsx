import React from 'react'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCircleNotch } from '@fortawesome/free-solid-svg-icons'

import '../../../../static/App.css'
import Style from '../../../../styles/components/pageAdmin.module.css'

import { fetchUsers } from '../../../../actions/index'

export class UserList extends React.Component<any, any> {
    constructor(props: any) {
        super(props)
        this.state = {
            customers_fetched: false,
        }
    }

    // intervalID var to keep track of auto-refreshing across functions
    intervalID: any | undefined

    // TODO (adriano) shouldn't we only fetch customers if we have an access token?
    componentDidMount() {
        // refresh the customer table every 60 seconds
        this.intervalID = setInterval(this.getUpdatedDatabase.bind(this), 60000)
        this.props.dispatch(fetchUsers())
    }

    componentWillUnmount() {
        // stop auto-refreshing
        clearInterval(this.intervalID)
    }

    componentDidUpdate() {
        if (
            this.props.access_token &&
            this.props.customers.length === 0 &&
            !this.state.customers_fetched
        ) {
            this.props.dispatch(fetchUsers())
            this.setState({ customers_fetched: true })

            // refresh the customer table every 60 seconds
            this.intervalID = setInterval(
                this.getUpdatedDatabase.bind(this),
                60000
            )
        }
    }

    getUpdatedDatabase() {
        this.props.dispatch(fetchUsers())
    }

    render() {
        let component = this

        return (
            <div
                style={{
                    maxHeight: 300,
                    overflowY: 'scroll',
                }}
            >
                {this.props.customers.length > 0 ? (
                    <div
                        style={{
                            display: 'flex',
                            flexWrap: 'wrap',
                        }}
                    >
                        {this.props.customers.map(function (
                            value: any,
                            index: any
                        ) {
                            if (value && value.name && value.name !== '') {
                                return (
                                    <div
                                        key={index}
                                        className="pointerOnHover"
                                        onClick={() => {
                                            component.props.openModal(
                                                value.name
                                            )
                                        }}
                                        style={{
                                            padding: '5px 10px',
                                            marginRight: 10,
                                            marginBottom: 10,
                                            fontSize: 13,
                                            color: '#8884d8',
                                            background:
                                                'rgba(136, 132, 216, 0.1)',
                                        }}
                                    >
                                        {value.name}
                                    </div>
                                )
                            } else {
                                return <div key={index}></div>
                            }
                        })}
                    </div>
                ) : (
                    <div className={Style.spinnerContainer}>
                        <div style={{ width: '100%', textAlign: 'center' }}>
                            <FontAwesomeIcon
                                icon={faCircleNotch}
                                spin
                                style={{
                                    color: '#1e1f36',
                                    margin: 'auto',
                                    marginTop: 220,
                                }}
                            />
                        </div>
                    </div>
                )}
            </div>
        )
    }
}

function mapStateToProps(state: any) {
    return {
        customers: state.AccountReducer.customers
            ? state.AccountReducer.customers.reverse()
            : [],
        access_token: state.AccountReducer.access_token,
    }
}

export default connect(mapStateToProps)(UserList)
