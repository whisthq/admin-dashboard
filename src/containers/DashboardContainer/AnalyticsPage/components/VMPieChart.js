import React, { Component } from 'react'
import { connect } from 'react-redux'
import { PieChart, Pie, ResponsiveContainer, Cell } from 'recharts'

import '../../../../static/App.css'

class VMPieChart extends Component {
    render() {
        const data = [
            { name: 'Deallocated', value: this.props.deallocated },
            { name: 'Running Available', value: this.props.available },
            { name: 'Running Unavailable', value: this.props.unavailable },
        ]
        const colors = ['#bf3762', '#15d157', '#e2e620']
        return (
            <div>
                <ResponsiveContainer width={200} height={200}>
                    <PieChart>
                        <text
                            x={100}
                            y={85}
                            textAnchor="middle"
                            dominantBaseline="middle"
                            style={{
                                fontSize: 40,
                                fontWeight: 'bold',
                            }}
                        >
                            {this.props.deallocated +
                                this.props.available +
                                this.props.unavailable}
                        </text>
                        <text
                            x={100}
                            y={125}
                            textAnchor="middle"
                            dominantBaseline="middle"
                            style={{
                                fontSize: 14,
                                fontWeight: 'normal',
                                color: '#777777',
                            }}
                        >
                            {this.props.location}
                        </text>
                        <Pie data={data} dataKey="value" innerRadius={70}>
                            {data.map((entry, index) => (
                                <Cell fill={colors[index % colors.length]} />
                            ))}
                        </Pie>
                    </PieChart>
                </ResponsiveContainer>
            </div>
        )
    }
}

export default connect()(VMPieChart)
