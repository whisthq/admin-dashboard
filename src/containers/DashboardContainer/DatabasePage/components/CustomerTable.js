import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Table } from 'antd'
import moment from 'moment'

import '../../../../static/App.css'
import 'antd/dist/antd.css'
import Style from '../../../../styles/components/pageAdmin.module.css'

import { fetchCustomers } from '../../../../actions/index'

class CustomerTable extends Component {
    constructor(props) {
        super(props)
        this.state = {
            customers_fetched: false,
            filterKeyword: '',
        }
    }

    // intervalID var to keep track of auto-refreshing across functions
    intervalID

    componentDidMount() {
        this.intervalID = setInterval(this.getUpdatedDatabase.bind(this), 60000)
    }

    componentWillUnmount() {
        // stop auto-refreshing
        clearInterval(this.intervalID)
    }

    componentDidUpdate(prevProps) {
        if (
            this.props.access_token &&
            this.props.customers.length === 0 &&
            !this.state.customers_fetched
        ) {
            this.props.dispatch(fetchCustomers())
            this.setState({ customers_fetched: true })

            // refresh the customer table every 60 seconds
            this.intervalID = setInterval(
                this.getUpdatedDatabase.bind(this),
                60000
            )
        }
    }

    getUpdatedDatabase() {
        this.props.dispatch(fetchCustomers())
    }

    keywordFilter = (e) => {
        let filterKeyword = e.target.value
        this.setState({
            filterKeyword: filterKeyword,
        })
    }

    render() {
        let columns = []
        let data = []
        if (this.props.customers && this.props.customers.length) {
            Object.keys(this.props.customers[0]).forEach(function (key) {
                let fixWidth = false
                if (key === 'username') {
                    fixWidth = 200
                } else if (key === 'trial_end' || key === 'created') {
                    fixWidth = 150
                }
                let customRender = false
                if (key === 'trial_end' || key === 'created') {
                    customRender = (text) => (
                        <p>{moment(text * 1000).format('lll')}</p>
                    )
                }
                columns.push({
                    title: key,
                    dataIndex: key,
                    sorter: (a, b) => {
                        if (a[key] === null) {
                            return 1
                        }
                        if (b[key] === null) {
                            return -1
                        }

                        var a_temp = a[key].toString().toLowerCase()
                        var b_temp = b[key].toString().toLowerCase()
                        if (a_temp > b_temp) {
                            return 1
                        } else if (a_temp < b_temp) {
                            return -1
                        }

                        return 0
                    },
                    width: fixWidth,
                    render: customRender,
                })
            })
            columns.reverse()

            this.props.customers.forEach(function (customer) {
                data.push(customer)
            })
            if (this.state.filterKeyword) {
                data = data.filter((element) => {
                    return element.username.includes(this.state.filterKeyword)
                })
            }
        }

        return (
            <div className={Style.tableWrapper}>
                <input
                    type="text"
                    placeholder="Filter by keyword"
                    style={{
                        width: '100%',
                        border: 'none',
                        padding: '10px 20px',
                        background: '#ebecf0',
                    }}
                    onChange={this.keywordFilter}
                />
                <Table
                    columns={columns}
                    dataSource={data}
                    scroll={{ x: 1000, y: 450 }}
                    size="middle"
                    rowClassName={Style.tableRow}
                    loading={this.props.customers.length === 0}
                />
            </div>
        )
    }
}

function mapStateToProps(state) {
    return {
        customers: state.AccountReducer.customers
            ? state.AccountReducer.customers.reverse()
            : [],
        access_token: state.AccountReducer.access_token,
    }
}

export default connect(mapStateToProps)(CustomerTable)
