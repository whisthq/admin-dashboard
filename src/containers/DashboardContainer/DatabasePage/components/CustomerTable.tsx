import React from 'react'
import { connect } from 'react-redux'
import { Table } from 'antd'
import moment from 'moment'

import '../../../../static/App.css'
import 'antd/dist/antd.css'
import Style from '../../../../styles/components/pageAdmin.module.css'

import { fetchCustomers } from '../../../../actions/index'

class CustomerTable extends React.Component<any, any> {
    constructor(props: any) {
        super(props)
        this.state = {
            customers_fetched: false,
            filterKeyword: '',
        }
    }

    // intervalID var to keep track of auto-refreshing across functions
    intervalID: any | undefined

    componentDidMount() {
        this.intervalID = setInterval(this.getUpdatedDatabase.bind(this), 60000)
    }

    componentWillUnmount() {
        // stop auto-refreshing
        clearInterval(this.intervalID)
    }

    componentDidUpdate() {
        if (
            this.props.access_token &&
            this.props.customers.length === 0 &&
            !this.state.customers_fetched
        ) {
            this.props.dispatch(fetchCustomers())
            this.setState({ customers_fetched: true })

            // refresh the customer table every 60 seconds
            this.intervalID = setInterval(
                this.getUpdatedDatabase.bind(this),
                60000
            )
        }
    }

    getUpdatedDatabase() {
        this.props.dispatch(fetchCustomers())
    }

    keywordFilter = (e: any) => {
        let filterKeyword = e.target.value
        this.setState({
            filterKeyword: filterKeyword,
        })
    }

    render() {
        let columns: any[] = []
        let data: any[] = [] 
        if (this.props.customers && this.props.customers.length) {
            Object.keys(this.props.customers[0]).forEach(function (key) {
                let fixWidth: any = false
                if (key === 'username') {
                    fixWidth = 200
                } else if (key === 'trial_end' || key === 'created') {
                    fixWidth = 150
                }
                let customRender: any = false
                if (key === 'trial_end' || key === 'created') {
                    customRender = (text: any) => (
                        <p>{moment(text * 1000).format('lll')}</p>
                    )
                }
                columns.push({
                    title: key,
                    dataIndex: key,
                    sorter: (a: any, b: any) => {
                        if (a[key] === null) {
                            return 1
                        }
                        if (b[key] === null) {
                            return -1
                        }

                        var a_temp = a[key].toString().toLowerCase()
                        var b_temp = b[key].toString().toLowerCase()
                        if (a_temp > b_temp) {
                            return 1
                        } else if (a_temp < b_temp) {
                            return -1
                        }

                        return 0
                    },
                    width: fixWidth,
                    render: customRender,
                })
            })
            columns.reverse()

            this.props.customers.forEach(function (customer: any) {
                data.push(customer)
            })
            if (this.state.filterKeyword) {
                data = data.filter((element) => {
                    return element.username.includes(this.state.filterKeyword)
                })
            }
        }

        return (
            <div className={Style.tableWrapper}>
                <input
                    type="text"
                    placeholder="Filter by keyword"
                    style={{
                        width: '100%',
                        border: 'none',
                        padding: '10px 20px',
                        background: '#ebecf0',
                    }}
                    onChange={this.keywordFilter}
                />
                <Table
                    columns={columns}
                    dataSource={data}
                    scroll={{ x: 1000, y: 450 }}
                    size="middle"
                    rowClassName={Style.tableRow}
                    loading={this.props.customers.length === 0}
                />
            </div>
        )
    }
}

function mapStateToProps(state: any) {
    return {
        customers: state.AccountReducer.customers
            ? state.AccountReducer.customers.reverse()
            : [],
        access_token: state.AccountReducer.access_token,
    }
}

export default connect(mapStateToProps)(CustomerTable)
