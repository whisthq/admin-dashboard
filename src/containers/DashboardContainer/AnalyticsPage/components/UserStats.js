import React, { Component } from 'react'
import { ToggleButtonGroup, ToggleButton } from 'react-bootstrap'
import { connect } from 'react-redux'
import moment from 'moment'
import {
    BarChart,
    Bar,
    ResponsiveContainer,
    XAxis,
    YAxis,
    Tooltip,
} from 'recharts'

import { fetchUserReport } from 'actions/index.js'

import CustomerList from 'containers/DashboardContainer/AnalyticsPage/components/CustomerList'

class UserStats extends Component {
    constructor(props) {
        super(props)
        this.state = {
            modalOpen: false,
            username: null,
            timescale: 'week',
            activity: null,
        }
    }

    componentDidUpdate(prevProps) {
        if (prevProps.userReport !== this.props.userReport) {
            let activity = []
            if (this.props.userReport.length) {
                this.props.userReport.forEach((element) => {
                    activity.push({
                        x: element.timestamp,
                        'Time online': element.minutes,
                    })
                })
            }

            this.setState({ activity: activity })
        }
    }

    openModal = (username) => {
        this.setState({ modalOpen: true, username: username })
        this.props.dispatch(fetchUserReport(this.state.timescale, username))
    }

    closeModal = () => {
        this.setState({ modalOpen: false, timescale: 'week' })
    }

    handleChartSelect = (val) => {
        console.log(val)
        this.setState({ timescale: val })
        this.props.dispatch(fetchUserReport(val, this.state.username))
    }

    render() {
        let format = 'MMM Do'

        return (
            <div>
                {this.state.activity && this.state.activity.length > 0 ? (
                    <div
                        style={{
                            position: 'relative',
                            right: 40,
                        }}
                    >
                        <div style={{ textAlign: 'right' }}>
                            <ToggleButtonGroup
                                type="radio"
                                name="select"
                                defaultValue={'week'}
                            >
                                {/* <ToggleButton value={'day'}>
                                24 Hours
                            </ToggleButton> */}
                                <ToggleButton
                                    value={'week'}
                                    onClick={() =>
                                        this.handleChartSelect('week')
                                    }
                                >
                                    7 Days
                                </ToggleButton>
                                <ToggleButton
                                    value={'month'}
                                    onClick={() =>
                                        this.handleChartSelect('month')
                                    }
                                >
                                    30 Days
                                </ToggleButton>
                                {/* <ToggleButton value={'all'}>
                                All Time
                            </ToggleButton> */}
                            </ToggleButtonGroup>
                        </div>
                        <ResponsiveContainer height={300} width="100%">
                            <BarChart data={this.state.activity}>
                                <XAxis
                                    dataKey="x"
                                    domain={
                                        this.state.timescale === 'week'
                                            ? [
                                                  moment()
                                                      .subtract('days', 8)
                                                      .unix(),
                                                  Date.now() / 1000,
                                              ]
                                            : [
                                                  moment()
                                                      .subtract('months', 1)
                                                      .unix(),
                                                  Date.now() / 1000,
                                              ]
                                    }
                                    tick={{
                                        fontSize: 12,
                                        transform: 'translate(0, 10)',
                                    }}
                                    tickFormatter={(unixTime) =>
                                        moment(unixTime * 1000).format(format)
                                    }
                                    type="number"
                                    scale="time"
                                    dx={20}
                                />
                                <YAxis
                                    dataKey="Time online"
                                    allowDecimals={false}
                                    tick={{
                                        fontSize: 12,
                                    }}
                                    axisLine={false}
                                />
                                <Tooltip
                                    contentStyle={{
                                        border: 'none',
                                        fontSize: 14,
                                    }}
                                    labelFormatter={(unixTime) =>
                                        moment(unixTime * 1000).format(format)
                                    }
                                />
                                <Bar
                                    dataKey="Time online"
                                    barSize={10}
                                    fill="#4636a6"
                                    formatter={(value, name, entry) =>
                                        Math.round(value / 60).toString() +
                                        'h ' +
                                        (value % 60).toString() +
                                        'm'
                                    }
                                />
                            </BarChart>
                        </ResponsiveContainer>
                    </div>
                ) : (
                    <div
                        style={{
                            width: '100%',
                            height: 300,
                            textAlign: 'center',
                            paddingTop: 140,
                            background: '#f5f7fa',
                            fontWeight: 'bold',
                            borderRadius: 4,
                        }}
                    >
                        {!this.state.username || this.state.username === '' ? (
                            <div>Click on a user to view their data</div>
                        ) : (
                            <div>
                                No data available for {this.state.username}
                            </div>
                        )}
                    </div>
                )}
                <div style={{ marginTop: 50 }}>
                    <CustomerList openModal={this.openModal} />
                </div>
            </div>
        )
    }
}

function mapStateToProps(state) {
    return {
        userReport: state.AccountReducer.userReport,
    }
}

export default connect(mapStateToProps)(UserStats)
