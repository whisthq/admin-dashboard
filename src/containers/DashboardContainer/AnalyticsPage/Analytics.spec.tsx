import React from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { Row, Col } from 'react-bootstrap'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'

import { Analytics } from './Analytics'

import { VMPieChart } from './components/VMPieChart'
import { UserStats } from './components/UserStats'
import { GeneralStats } from './components/GeneralStats'
import { SummaryStats } from './components/SummaryStats'

// necessary for enzyme to work
Enzyme.configure({ adapter: new Adapter() })

/*
The Analytics component renders the ANALYTICS page in the dashboard.

It must have two horizontally aligned components: on the left there is are tabs which display
sub-components and on the right there are three VM pie charts that display labels for the regions with the
total number of deallocated, unavailable, and available numbers of VMs passed as props. The order of
these pie charts does not matter but the correct label should correspond to its pie chart.
*/

describe('<Analytics />', () => {
    test('ensure page change was dispatched to redux on component mount', () => {
        // TODO
    })

    it('renders a row with two cols', () => {
        const wrapper = shallow(
            <Analytics
                dispatch={(items: any) => {
                    // mock
                }}
            />
        )
        expect(wrapper.find(Row)).toHaveLength(1)
        expect(wrapper.find(Col)).toHaveLength(2)
    })

    it('renders the three tabs with general stats, user stats, summary stats', () => {
        const wrapper = shallow(
            <Analytics
                dispatch={(items: any) => {
                    // mock
                }}
            />
        )
        expect(wrapper.find(Tabs)).toHaveLength(1)

        expect(wrapper.find(TabList)).toHaveLength(1)
        expect(wrapper.find(Tab)).toHaveLength(3)

        // each of the tab panel can be a loading screen inside the component, but not here
        expect(wrapper.find(TabPanel)).toHaveLength(3)

        expect(wrapper.find(UserStats)).toHaveLength(1)
        expect(wrapper.find(SummaryStats)).toHaveLength(1)
        expect(wrapper.find(GeneralStats)).toHaveLength(1)
    })

    it('renders three vm pie charts', () => {
        const wrapper = shallow(
            <Analytics
                dispatch={(items: any) => {
                    // mock
                }}
            />
        )
        expect(wrapper.find(VMPieChart)).toHaveLength(3)
    })

    it('renders the correct labels for the pie charts', () => {
        const labels = ['Eastus', 'Northcentralus', 'Southcentralus']
        const remainingLabels = new Set(labels)

        const wrapper = shallow(
            <Analytics
                dispatch={(items: any) => {
                    // mock
                }}
            />
        )
        wrapper.find(VMPieChart).forEach((vmPieChart) => {
            let location = vmPieChart.prop('location')
            expect(remainingLabels.has(location)).toEqual(true)
            remainingLabels.delete(location)
        })
    })
})
