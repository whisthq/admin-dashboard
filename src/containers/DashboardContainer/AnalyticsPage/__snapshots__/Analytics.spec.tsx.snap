// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Analytics /> renders correctly 1`] = `
<div
  style={
    Object {
      "bottom": 10,
      "position": "relative",
      "width": "100vw",
    }
  }
>
  <div
    style={
      Object {
        "color": "#111111",
        "fontSize": 45,
        "fontWeight": "bold",
        "marginBottom": 30,
      }
    }
  >
    ANALYTICS
  </div>
  <Row
    noGutters={false}
  >
    <Col
      lg={8}
    >
      <Tabs
        defaultFocus={false}
        defaultIndex={null}
        forceRenderTabPanel={false}
        selectedIndex={null}
      >
        <TabList
          className="react-tabs__tab-list"
          style={
            Object {
              "border": "none",
              "fontSize": 16,
              "marginBottom": 20,
              "textAlign": "left",
            }
          }
        >
          <Tab
            className="react-tabs__tab"
            disabledClassName="react-tabs__tab--disabled"
            focus={false}
            id={null}
            panelId={null}
            selected={false}
            selectedClassName="react-tabs__tab--selected"
          >
            General
          </Tab>
          <Tab
            className="react-tabs__tab"
            disabledClassName="react-tabs__tab--disabled"
            focus={false}
            id={null}
            panelId={null}
            selected={false}
            selectedClassName="react-tabs__tab--selected"
          >
            Users
          </Tab>
          <Tab
            className="react-tabs__tab"
            disabledClassName="react-tabs__tab--disabled"
            focus={false}
            id={null}
            panelId={null}
            selected={false}
            selectedClassName="react-tabs__tab--selected"
          >
            Stats
          </Tab>
        </TabList>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          selectedClassName="react-tabs__tab-panel--selected"
        >
          <Connect(GeneralStats) />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          selectedClassName="react-tabs__tab-panel--selected"
        >
          <Connect(UserStats) />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          selectedClassName="react-tabs__tab-panel--selected"
        >
          <Connect(SummaryStats) />
        </TabPanel>
      </Tabs>
    </Col>
    <Col>
      <div
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 30,
            "marginTop": 60,
          }
        }
      >
        Cloud PCs
      </div>
      <Connect(VMPieChart)
        available={0}
        deallocated={0}
        location="Eastus"
        unavailable={0}
      />
      <Connect(VMPieChart)
        available={0}
        deallocated={0}
        location="Northcentralus"
        unavailable={0}
      />
      <Connect(VMPieChart)
        available={0}
        deallocated={0}
        location="Southcentralus"
        unavailable={0}
      />
    </Col>
  </Row>
</div>
`;
